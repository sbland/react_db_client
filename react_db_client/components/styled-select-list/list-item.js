import React from 'react'
import PropTypes from 'prop-types'
import { stringifyData } from '@samnbuk/react_db_client.helpers.data-processing';

import styles from './StyledSelectList.module.scss'

export const ListItem = ({
  data,
  handleSelect,
  headings,
  columnWidths,
  customParsers,
  isSelected,
  classes,
}) => {
  return (
    <button
      key={data.uid}
      className={`${styles.styledList_itemBtn} ${classes.styledList_itemBtn} ${isSelected ? classes.styledListItem_item_selected : ''}`}
      type='button'
      onClick={() => handleSelect(data.uid, data)}
      role='listitem'
    >
      {headings.map((heading, i) => (
        <div
          key={heading.uid}
          style={{
            width: columnWidths[i]
          }}
          className={`${styles.styledList_itemCell} ${
            isSelected ? 'selected' : ''
          }`}
        >
          {stringifyData(data[heading.uid], heading, customParsers)}
        </div>
      ))}
    </button>
  )
}

ListItem.propTypes = {
  data: PropTypes.shape({
    uid: PropTypes.string.isRequired
  }).isRequired,
  headings: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired
    })
  ).isRequired,
  handleSelect: PropTypes.func.isRequired,
  customParsers: PropTypes.objectOf(PropTypes.func).isRequired,
  columnWidths: PropTypes.arrayOf(PropTypes.number).isRequired,
  isSelected: PropTypes.bool.isRequired,
  classes: PropTypes.objectOf(PropTypes.string),
};

ListItem.defaultProps = {
    classes: styles,
}

