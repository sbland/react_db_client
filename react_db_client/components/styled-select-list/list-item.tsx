import React from 'react';
import PropTypes from 'prop-types';
import { stringifyData } from '@react_db_client/helpers.data-processing';
import { Uid } from '@react_db_client/constants.client-types';

import {
  StyledListItemBtnStyle,
  StyledListItemCell,
  StyledListItemStyle,
} from './styles';
import { IHeading, IItem } from './lib';

export interface IListItemProps<ItemType extends IItem> {
  data: ItemType;
  handleSelect: (uid: Uid, data: ItemType) => void;
  headings: IHeading[];
  columnWidths: number[];
  tableWidth: number;
  customParsers: { [k: string]: (valIn: any) => any };
  isSelected?: boolean;
}

export const ListItem = <ItemType extends IItem>({
  data,
  handleSelect,
  headings,
  columnWidths,
  tableWidth,
  customParsers,
  isSelected,
}: IListItemProps<ItemType>) => {
  return (
    <StyledListItemStyle style={{ width: tableWidth }}>
      <StyledListItemBtnStyle
        key={data.uid}
        type="button"
        onClick={() => handleSelect(data.uid, data)}
        style={{ width: tableWidth }}
        className={`styledList_itemBtn ${
          isSelected ? 'styledList_itemBtnSelected' : ''
        }`}
        selected={isSelected}
        data-isselected={isSelected}
        aria-label={data.label}
      >
        {headings.map((heading, i) => (
          <StyledListItemCell
            key={heading.uid}
            data-testid={`SaS_item_cell_${heading.uid}`}
            style={{
              width: columnWidths[i],
            }}
          >
            {stringifyData(data[heading.uid], heading, customParsers)}
          </StyledListItemCell>
        ))}
      </StyledListItemBtnStyle>
    </StyledListItemStyle>
  );
};

ListItem.propTypes = {
  data: PropTypes.shape({
    uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  }).isRequired,
  headings: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
    })
  ).isRequired,
  handleSelect: PropTypes.func.isRequired,
  customParsers: PropTypes.objectOf(PropTypes.func).isRequired,
  columnWidths: PropTypes.arrayOf(PropTypes.number).isRequired,
  isSelected: PropTypes.bool,
};

ListItem.defaultProps = {
  isSelected: false,
};
