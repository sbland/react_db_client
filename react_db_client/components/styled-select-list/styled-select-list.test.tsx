import React from 'react';
import { screen, render, within } from '@testing-library/react';
import UserEvent from '@testing-library/user-event';

import * as compositions from './list-item.composition';

describe('Styled select list', () => {
  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, async () => {
        render(<Composition />);
        // @ts-ignore
        if (Composition.waitForReady) await Composition.waitForReady();
      });
    });
  });
});
