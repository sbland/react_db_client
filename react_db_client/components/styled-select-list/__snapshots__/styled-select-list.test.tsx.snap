// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styled select list Snapshot 1`] = `
<div
  className="sc-gswNZR leMfon"
  style={Object {}}
>
  <div
    className="sc-dkrFOg iujzfJ"
  >
    <div
      className="sc-jSUZER kdPRE"
      style={
        Object {
          "width": 360,
        }
      }
    >
      UID
    </div>
    <div
      className="sc-jSUZER kdPRE"
      style={
        Object {
          "width": 360,
        }
      }
    >
      Type
    </div>
    <div
      className="sc-jSUZER kdPRE"
      style={
        Object {
          "width": 360,
        }
      }
    >
      Name
    </div>
  </div>
  <ul
    className="sc-hLBbgP lkHHzI"
    role="list"
    style={
      Object {
        "zIndex": 10,
      }
    }
  >
    <button
      className="sc-eDvSVe jYBAgP"
      onClick={[Function]}
      role="listitem"
      type="button"
    >
      <div
        className="sc-gKPRtg lyePn"
        selected={false}
        style={
          Object {
            "width": 360,
          }
        }
      >
        filterA
      </div>
      <div
        className="sc-gKPRtg lyePn"
        selected={false}
        style={
          Object {
            "width": 360,
          }
        }
      >
        text
      </div>
      <div
        className="sc-gKPRtg lyePn"
        selected={false}
        style={
          Object {
            "width": 360,
          }
        }
      >
        
      </div>
    </button>
    <button
      className="sc-eDvSVe jYBAgP"
      onClick={[Function]}
      role="listitem"
      type="button"
    >
      <div
        className="sc-gKPRtg lyePn"
        selected={false}
        style={
          Object {
            "width": 360,
          }
        }
      >
        filterB
      </div>
      <div
        className="sc-gKPRtg lyePn"
        selected={false}
        style={
          Object {
            "width": 360,
          }
        }
      >
        text
      </div>
      <div
        className="sc-gKPRtg lyePn"
        selected={false}
        style={
          Object {
            "width": 360,
          }
        }
      >
        
      </div>
    </button>
  </ul>
</div>
`;
